File: DSA Busted/Heap/HeapImplementation.cpp
#include<iostream>
#include<queue>
using namespace std;
class Heap{
    public:
        int arr[100];
        int size;
        Heap(){
            size= 0;
        }
        void insert(int val){
            size++;
            int index= size;
            arr[size]= val;
            while(index>1){
                if(arr[index]>arr[index/2]){
                    swap(arr[index], arr[index/2]);
                    index= index/2;
                }else{
                    return;
                }
            }
        }
        void deleteFromHeap(){
            if (size==0)
            {
                cout<<"Heap is empty"<<endl;
                return;
            }
            
            arr[1]= arr[size];
            size--;

            int i= 1;
            while(i<size){
                int left= 2*i;
                int right= 2*i+1;
                if(left<=size && arr[left]>arr[1]){
                    swap(arr[left], arr[1]);
                    i= left;
                }
                else if(right<=size && arr[right]>arr[1]){
                    swap(arr[right], arr[1]);
                    i= right;
                }
                else{
                    return;
                }
            }

        }
        void print(){
            cout<<endl;
            for (int i=1; i<=size; i++)
            {
                cout<<arr[i]<<" ";
            }
            cout<<endl;
        }
};
int main(){
    Heap heap;
    heap.insert(55);
    heap.insert(54);
    heap.insert(53);
    heap.insert(50);
    heap.insert(52);
    heap.print();
    heap.deleteFromHeap();
    heap.print();

    // Max Heap
    priority_queue<int> maxHeap;
    maxHeap.push(1);
    maxHeap.push(3);
    maxHeap.push(6);
    maxHeap.push(4);
    maxHeap.push(2);
    maxHeap.push(21);

    cout<<"Max Heap "<<endl;
    while (!maxHeap.empty())
    {
        cout<<maxHeap.top()<<" ";
        maxHeap.pop();
    }

    cout<<endl;

    priority_queue<int, vector<int>, greater<int>> minHeap;
    minHeap.push(1);
    minHeap.push(3);
    minHeap.push(6);
    minHeap.push(4);
    minHeap.push(2);
    minHeap.push(21);
    

    cout<<"Min Heap "<<endl;
    while (!minHeap.empty())
    {
        cout<<minHeap.top()<<" ";
        minHeap.pop();
    }

    cout<<endl;


    return 0;
}

File: DSA-Problems/Heap/HeapImplementation.java
import java.util.*;

class HeapImplementation {
    int[] arr = new int[101]; // Array to store heap elements (1-based indexing)
    int size;

    public HeapImplementation() {
        size = 0; // Initially the heap is empty
    }

    // Insert a new value into the heap
    public void insert(int val) {
        size++;
        int index = size;
        arr[index] = val;

        // Heapify Up: Restore max-heap property by comparing with parent
        while (index > 1) {
            int parent = index / 2;
            if (arr[index] > arr[parent]) {
                // Swap if current is greater than parent
                int temp = arr[index];
                arr[index] = arr[parent];
                arr[parent] = temp;

                index = parent; // Move up
            } else {
                return; // Heap property is satisfied
            }
        }
    }

    // Delete the root (maximum element) from the heap
    public void deleteFromHeap() {
        if (size == 0) {
            System.out.println("Heap is empty");
            return;
        }

        // Replace root with last element and reduce size
        arr[1] = arr[size];
        size--;

        // Heapify Down: Restore heap property
        int i = 1;
        while (i <= size) {
            int left = 2 * i;
            int right = 2 * i + 1;
            int largest = i;

            if (left <= size && arr[left] > arr[largest]) {
                largest = left;
            }
            if (right <= size && arr[right] > arr[largest]) {
                largest = right;
            }

            if (largest != i) {
                // Swap current with largest child
                int temp = arr[i];
                arr[i] = arr[largest];
                arr[largest] = temp;
                i = largest; // Move down
            } else {
                return; // Heap property satisfied
            }
        }
    }

    // Print all elements in the heap
    public void print() {
        System.out.println();
        for (int i = 1; i <= size; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
}

class Main {
    public static void main(String[] args) {
        // Manual Max Heap
        HeapImplementation heap = new HeapImplementation();
        heap.insert(55);
        heap.insert(54);
        heap.insert(53);
        heap.insert(50);
        heap.insert(52);
        heap.print();

        heap.deleteFromHeap();
        heap.print();

        // Max Heap using Java's PriorityQueue (default is Min Heap, so we use reverse order)
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());
        maxHeap.add(1);
        maxHeap.add(3);
        maxHeap.add(6);
        maxHeap.add(4);
        maxHeap.add(2);
        maxHeap.add(21);

        System.out.println("Max Heap (PriorityQueue):");
        while (!maxHeap.isEmpty()) {
            System.out.print(maxHeap.poll() + " ");
        }
        System.out.println();

        // Min Heap using Java's PriorityQueue (default behavior)
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        minHeap.add(1);
        minHeap.add(3);
        minHeap.add(6);
        minHeap.add(4);
        minHeap.add(2);
        minHeap.add(21);

        System.out.println("Min Heap (PriorityQueue):");
        while (!minHeap.isEmpty()) {
            System.out.print(minHeap.poll() + " ");
        }
        System.out.println();
    }
}
